src/asmcnc/comms/grbl_settings_manager.py

```diff
@@ -149,3 +236,56 @@ def on_setting_steps_per_mm(self, instance, value, setting):
                 self.save_machine_data_to_file()
                 log('Calibration run? Updated {} in file: {}'.
                     format(descriptions[setting], value))
+
+    def on_persistent_setting(self, instance, value, setting):
+        """
+        Called when a persistent setting is read in.
+
+        Checks for default value and restores value from file if possible.
+        """
+        descriptions = {
+            0: "Step pulse, microseconds ($0) setting",
+            1: "Step idle delay, milliseco ($1) setting",
+            2: "Step port invert, mask ($2) setting",
+            4: "Step enable invert, boolean ($4) setting",
+            5: "Limit pins invert, boolean ($5) setting",
+            6: "Probe pin invert, boolean ($6) setting",
+            10: "Status report, mask ($10) setting",
+            11: "Junction deviation, mm ($11) setting",
+            12: "Arc tolerance, mm ($12) setting",
+            13: "Report inches, boolean ($13) setting",
+            20: "Soft limits, boolean ($20) setting",
+            21: "Hard limits, boolean ($21) setting",
+            22: "Homing cycle, boolean ($22) setting",
+            23: "Homing dir invert, mask ($23) setting",
+            24: "Homing feed, mm/min ($24) setting",
+            25: "Homing seek, mm/min ($25) setting",
+            26: "Homing debounce, millisec ($26) setting",
+            27: "Homing pull-off, mm ($27) setting",
+            30: "Max spindle speed, RPM ($30) setting",
+            31: "Min spindle speed, RPM ($31) setting",
+            32: "Laser mode, boolean ($32) setting",
+            110: "X Max rate, mm/min ($110) setting",
+            111: "Y Max rate, mm/min ($111) setting",
+            112: "Z Max rate, mm/min ($112) setting",
+            120: "X Acceleration, mm/s ($120) setting",
+            121: "Y Acceleration, mm/s ($121) setting",
+            122: "Z Acceleration, mm/s ($122) setting",
+            130: "X Max travel, mm ($130) setting"
+        }
+
+        if self._machine_saved_data[setting] is None:  # not saved yet?
```

Comment:
You don't need to save the machine data from zh to file. Just check if the zh-value deviates from the default setting, if so write the correct value from _system_default_data back to the zh.